eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceClientInstantiationCheck.java

```diff
@@ -0,0 +1,220 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.tools.checkstyle.checks;
+
+import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+import com.puppycrawl.tools.checkstyle.checks.naming.AccessModifier;
+import com.puppycrawl.tools.checkstyle.utils.CheckUtil;
+
+/**
+ * Verify the classes with annotation @ServiceClient should have following rules:
+ * <0l>
+ *   <li>No public or protected constructors</li>
+ *   <li>No public static method named 'builder'</li>
+ *   <li>Since these classes are supposed to be immutable, all fields in the service client classes should be final.</li>
+ * </0l>
+ */
+public class ServiceClientInstantiationCheck extends AbstractCheck {
+    private static final String SERVICE_CLIENT = "ServiceClient";
+    private static final String BUILDER = "builder";
+    private static final String ASYNC_CLIENT ="AsyncClient";
+    private static final String CLIENT = "Client";
+    private static final String IS_ASYNC = "isAsync";
+
+    private static boolean hasServiceClientAnnotation;
+    private static boolean isAsync;
+
+    @Override
+    public int[] getDefaultTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return new int[] {
+            TokenTypes.CLASS_DEF,
+            TokenTypes.CTOR_DEF,
+            TokenTypes.METHOD_DEF,
+            TokenTypes.OBJBLOCK
+        };
+    }
+
+    @Override
+    public void beginTree(DetailAST root) {
+        hasServiceClientAnnotation = true;
+        isAsync = false;
+    }
+
+    @Override
+    public void visitToken(DetailAST token) {
+        if (!hasServiceClientAnnotation) {
+            return;
+        }
+
+        switch (token.getType()) {
+            case TokenTypes.CLASS_DEF:
+                hasServiceClientAnnotation = hasServiceClientAnnotation(token);
+                checkServiceClientNaming(token);
+                break;
+            case TokenTypes.CTOR_DEF:
+                checkConstructor(token);
```

Comment:
I would add pre-emptive checks on these constructor and method_def checks that:
```java
if (hasServiceClientAnnotation) {
   // then do the check. Otherwise, you're doing the check on classes that aren't serviceclients.
}
```

Reply:
The condition is already checked in at beginning of visitToken()

Reply:
Seems odd you would pre-emptively set the field to "true" when beginning a walk because you don't really know if that file has a ServiceClient annotation, yet.

Reply:
Setting 'true' value to be able to walk the tree, The DFS tree traversal will first visit CLASS_DEF. If the class is annotated with @ServiceClient, the hasServiceClientAnnotation() method will return true, otherwise, it will return false.

Reply:
I have made a wrong assumption. I thought one tree traversal will only have one CLASS_DEF. But it could have a nest inner class.